### PR TITLE
Get letter status from Gov Notify

### PR DESCRIPTION
**Pull request by mrosel**
Polls Gov notify for updates on letter status and updates the Document object based on that state.

### CODE REVIEW

File: app/jobs/hackney/income/jobs/save_and_send_letter_job.rb [hunk 1]
```diff
@@ -14,6 +14,7 @@ def perform(letter_html:, bucket_name:, filename:, document_id:)
           @document_id = document_id
           binary_letter_content = generate_pdf_binary(letter_html, document.uuid)
 
+          Rails.logger.info "uploading #{filename} to #{bucket_name}"
```
Comment (elena-vi) [inline app/jobs/hackney/income/jobs/save_and_send_letter_job.rb]: Maybe also show the doc id for if something goes wrong and we need to hunt down that upload
```suggestion
          Rails.logger.info "uploading document: #{document_id} - #{filename} to #{bucket_name}"
```

File: config/environments/development.rb [hunk 2]
```diff
@@ -12,19 +12,7 @@
...
-    config.cache_store = :memory_store
-    config.public_file_server.headers = {
-      'Cache-Control' => "public, max-age=#{2.days.seconds.to_i}"
-    }
-  else
-    config.action_controller.perform_caching = false
-
-    config.cache_store = :null_store
-  end
+  config.cache_store = :memory_store
```
Comment (elena-vi) [inline config/environments/development.rb]: no `redis_store`?

File: lib/hackney/notification/request_precompiled_letter_state.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+          notification_gateway.precompiled_letter_state(
+            message_id: message_id
+          )
+
+        update_document(message_id: message_id, status: response[:status])
+        response
+      end
+
+      def update_document(message_id:, status:)
+        # document_store = Hackney::Cloud::Document
```
Comment (elena-vi) [inline lib/hackney/notification/request_precompiled_letter_state.rb]: I think you might need this here..
```suggestion
        document_store = Hackney::Cloud::Document
```

Comment (alexdesi) [inline lib/hackney/notification/request_precompiled_letter_state.rb]: unnecessary comment?

File: lib/hackney/notification/request_precompiled_letter_state.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        response =
+          notification_gateway.precompiled_letter_state(
+            message_id: message_id
+          )
+
+        update_document(message_id: message_id, status: response[:status])
+        response
+      end
+
+      def update_document(message_id:, status:)
```
Comment (alexdesi) [inline lib/hackney/notification/request_precompiled_letter_state.rb]: I think this should be private

File: app/jobs/hackney/income/jobs/save_and_send_letter_job.rb (hunk 1 above)
Comment (mrosel) [reply app/jobs/hackney/income/jobs/save_and_send_letter_job.rb]: :+1:

File: config/environments/development.rb (hunk 2 above)
Comment (mrosel) [reply config/environments/development.rb]: :thinking: the behavior is identical without the explicit need to run a separate redis server. It 'felt' like unnecessary overhead.

File: spec/hackney/income/jobs/save_and_send_letter_job_spec.rb [hunk 5]
```diff
@@ -17,14 +17,17 @@
...
+  let(:message_receipt) { Hackney::Notification::Domain::NotificationReceipt.new(body: 'body', message_id: SecureRandom.uuid) }
 
   before {
     expect_any_instance_of(Aws::S3::Encryption::Client).to receive(:put_object).and_return(AwsEncryptionClientDouble.new(nil).send(:put_object))
   }
 
   it 'uploads to clouds' do
     expect_any_instance_of(Aws::S3::Encryption::Client).to receive(:get_object).and_return(AwsClientResponse.new)
-    expect_any_instance_of(Hackney::Notification::GovNotifyGateway).to receive(:send_precompiled_letter).once
+    # expect_any_instance_of(Hackney::Notification::GovNotifyGateway).to receive(:send_precompiled_letter).once
```
Comment (alexdesi) [inline spec/hackney/income/jobs/save_and_send_letter_job_spec.rb]: unnecessary comments?

File: config/environments/development.rb (hunk 2 above)
Comment (elena-vi) [reply config/environments/development.rb]: okay cool

File: spec/support/stubs/stub_notifications_gateway.rb [hunk 6]
```diff
@@ -69,3 +70,37 @@ def get_template(id)
...
   end
 end
+
+# rubocop:disable Naming/VariableNumber
+# <Notifications::Client::Notification:0x000055faabc8b008
+class StubNotification
+  attr_reader :status
+  def initialize(reference:, id:)
+    @body = '',
+            @completed_at = Time.now,
```
Comment (alexdesi) [inline spec/support/stubs/stub_notifications_gateway.rb]: The intentation looks wrong here.

File: spec/hackney/income/jobs/save_and_send_letter_job_spec.rb [hunk 7]
```diff
@@ -40,7 +43,7 @@
   it 'creates pdf' do
     # expect_any_instance_of(Aws::S3::Encryption::Client).to receive(:put_object).and_return(AwsEncryptionClientDouble.new(nil).send(:put_object))
```
Comment (alexdesi) [inline spec/hackney/income/jobs/save_and_send_letter_job_spec.rb]: can you remove this in the meantime :) (I know this is out of scope)

**Pull request merged**